### PR TITLE
IndexedDB + caching

### PR DESCRIPTION
**Pull request by Lurrobert**
# Indexed DB function

- Save from EmbeddingIndex to IndexedDB
- Create and Delete DBs & objects stores
- search functionality

# Interface for testing IndexedDB

- generate embddings
- search
- delete
- view
image

### CODE REVIEW

**Comment by Lurrobert** [comment]
**please remember to use squash 😄  **

File: README.md [hunk 1]
```diff
@@ -57,54 +59,134 @@ We'll initially keep things super simple and sub 100ms
...
+```
+> **Note**: Both embeddings should be of the same length.
+
+---
+
+### Step 3: Create an Index
+Create an index with an initial array of objects. Each object must have an 'embedding' property.
+
+```ts
+const initialObjects = [...];
```
Comment (yusufhilmi) [inline README.md:88]: We should have 2,3 objects as an example.

File: README.md [hunk 2]
```diff
@@ -57,54 +59,134 @@ We'll initially keep things super simple and sub 100ms
...
+index.printIndex();
+```
+
+---
+
+### Step 10: Save to Database
+Save the index to a persistent IndexedDB database.
+
+```ts
+await index.saveIndexToDB("dbName", "ObjectStoreName");
```
Comment (yusufhilmi) [inline README.md:155]: saveIndexToDB is bogus, we need to specify IndexedDB and we prolly want a `saveIndex` method that generalizes for indexedDB, localStorage, a json on Node, and other cloud options starting from v0.1

File: README.md [hunk 3]
```diff
@@ -57,54 +59,134 @@ We'll initially keep things super simple and sub 100ms
 npm install client-vector-search
 ```
 
-## Usage
+## Usage Guide
```
Comment (yusufhilmi) [inline README.md:62]: I'm not a fan of these steps, we want copy-paste/able code that people can quickly try. We should have a `## Quickstart` and then this guide which should serve as Reference docs for all types, params, and other details.

File: package.json [hunk 4]
```diff
@@ -33,6 +34,8 @@
     "typescript": "^4.9.5"
   },
   "dependencies": {
-    "@xenova/transformers": "^2.5.2"
+    "@xenova/transformers": "^2.5.2",
+    "lru-cache": "^10.0.1",
+    "fake-indexeddb": "^4.0.2"
```
Comment (yusufhilmi) [inline package.json:39]: I'm not a fan of installing fake-indexeddb when we only use it for testing purposes.

File: README.md [hunk 5]
```diff
@@ -57,54 +59,134 @@ We'll initially keep things super simple and sub 100ms
...
+const results = await index.search(queryEmbedding, { topK: 5 }, true, 'dbName', 'ObjectStoreName');
+```
+
+---
+
+### Delete Database
+To delete an entire database.
+
+```ts
+await IndexedDbManager.deleteDB("dbName");
```
Comment (yusufhilmi) [inline README.md:173]: the indexedDBManager comes out of nowhere. We can and should store the details about indexedDB that is created from the index to index object itself with the next version for indexes with option: `{ indexedDB: true }`
```js
{    
     ...,
     indexedDB: {
       databases: [{ name: "clientVectorDB", objectStores: []}, ...] // name of the databases,
     }
 }
```
and have it as 
```js
const index = new EmbeddingIndex(initialObjects); 
await index.deleteIndexedDB("clientVectorDB")
```

File: README.md [hunk 6]
```diff
@@ -57,54 +59,134 @@ We'll initially keep things super simple and sub 100ms
...
-
-  // Step 8: Search the index with a query embedding
-  // Be careful: The query should be an embedding of type number[]
-  const queryEmbedding = await getEmbedding('Fruit'); // Query embedding
-  const results = index.search(queryEmbedding, { topK: 5 }); // Returns top similar objects
-
-  // Step 9: Print the entire index
-  // Be careful: This will print all objects in the index, which might be a large output for large indexes
-  index.printIndex(); // Prints the content of the index
+const allObjects = await IndexedDbManager.getAllObjectsFromDB("dbName", "ObjectStoreName");
```
Comment (yusufhilmi) [inline README.md:191]: we should add 
```js
const newIndex = new EmbeddingIndex(allObjects)
```

File: src/index.ts [hunk 7]
```diff
@@ -171,4 +195,89 @@ export class EmbeddingIndex {
       console.log(`Item ${idx + 1}:`, obj);
     });
   }
+
+  async saveIndexToDB(DBname: string = 'defaultDB', objectStoreName: string = 'DefaultStore'): Promise<void> {
```
Comment (yusufhilmi) [inline src/index.ts:199]: As I mentioned in the Readme, `saveIndex("indexedDB", { DBName: "clientVectorDB", objectStoreName:"ClientEmbeddings"})` is a nicer approach to this

File: test/README.md [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+## Cache files
+```cd test```
+
+```npx ts-node --esm src/cache.ts```
+
+## IndexedDB
+
+### test script
+
+1. uncomment fake DB linesin the `src/index.ts` library
```
Comment (yusufhilmi) [inline test/README.md:16]: We'll have a proper testing implemented once I resolve the issue with the jest. and hopefully get rid of this kinds of tricks completely including the whole test folder.

**Review by yusufhilmi — CHANGES_REQUESTED** [review]

File: test/src/indexed-db-front/package-lock.json
Comment (yusufhilmi) [inline test/src/indexed-db-front/package-lock.json:1]: we should ignore this file

**Pull request merged**